week-1-ts-ecosystem/w01e04-light-me-up/src/main.ts

```diff
@@ -8,14 +8,82 @@
  * Dodatkowo — rzucać błąd, jeśli zapas mocy się wyczerpie.
  * */
 
+const MILLISECONDS_IN_SECOND = 1000;
+
 class PowerSource {
-  private energySupply = 100
+  private energySupply = 100 // J
+
+  public consume(energy: number) {
+    const rest = this.energySupply - energy;
+
+    if (rest < 0) {
+      throw Error("Not enough energy");
+    }
 
-  consume(energy) {
-    this.energySupply -= energy
+    console.log(`Consume ${energy}J. Remaining energy: ` + rest + "J");
+    this.energySupply = rest;
   }
 }
 
 class LightBulb {
-  protected readonly powerConsumption = 20
+  protected readonly powerConsumption = 20 // W
+  private powerSource: PowerSource;
+  private consumption: number = -1;
+
+  constructor(powerSource: PowerSource) {
+    this.powerSource = powerSource;
+  }
+
+  consume () {
+    try {
+      this.powerSource.consume(this.powerConsumption);
+    } catch (error) {
+      console.log("Not enough energy. Switching off...");
+      this.switchOff();
+    }
+  }
+
+  switchOff () {
+      clearInterval(this.consumption);
+  }
+
+  switchOn (seconds: number) {
+      let secondsRemain = seconds;
+
```

Comment:
Tutaj brakuje mi wywołania:
```
this.switchOff()
```
Bo bardzo ładnie przygotowałeś tą metodę do wyłączenia żarówki - jako wyczyszczenie interwału.
Dzięki temu zapobiegasz ew. _memory leak_'om.

I tutaj jest taki moment na zabezpieczenie - żeby nie "zaświecić tej samej żarówki 2 razy" :)
to uruchomienie tego `switchOff()` by przed tym nas chroniło - można to tak wykorzystać.